### PR TITLE
feat: execute the GTFS Validator after downloading a dataset

### LINKED ISSUE #339: Implementation: Execute the GTFS Validator after downloading a dataset
### Describe the problem

Related to #228 
As part of the Datasets metadata, we would like to include a summary of the information produced by the [GTFS validator](https://github.com/MobilityData/gtfs-validator). This issue implements a process to execute the validator after downloading a dataset.

### Proposed solution

Tasks:
- [x] Implement execution of validator using GCP Workflow. POC available [here](https://console.cloud.google.com/workflows/workflow/northamerica-northeast1/test-workflow/source?project=mobility-feeds-dev)
- [x] Create Terraform script for deployment and EventArc trigger

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**
Closes #339 
The changes implemented in this PR are designed to automate the validation process, ensuring that the datasets folders are consistently augmented with the GTFS validation report. This automation is achieved through the introduction of a GCP Workflow that triggers the validator post-dataset download using the GTFS Validator's web services.

**Expected behavior:** 
From `mobility-feeds-dev` project:
- [HTML report example for dataset `mdb-742-202403212257`](https://storage.googleapis.com/mobilitydata-datasets-dev/mdb-742/mdb-742-202403212257/report.html)
- [JSON report example for dataset `mdb-742-202403212257`](https://storage.googleapis.com/mobilitydata-datasets-dev/mdb-742/mdb-742-202403212257/report.json)
- Execution logs:
Screenshot 2024-03-22 at 3 14 57 PM

**Testing tips:**
- **Using GCP Workflow execution**: 
  - Open the workflow `gtfs_validator_execution` in the GCP console ([here](https://console.cloud.google.com/workflows/workflow/northamerica-northeast1/gtfs_validator_execution/executions?project=mobility-feeds-dev))
  - Click `Execute`
  - Click `Use last input`
  - You can play with the `data.resourceName` field to specify a dataset or use the value provided. If you change this value make sure the dataset exists first.
  - Click `Execute`
  - Validate execution and results

Testers are invited to follow the tips AND to try anything they deem relevant outside the bounds of the testing tips. 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [ ] Add or update any needed documentation to the repo
- [ ] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: workflows/gtfs_validator_execution.yml [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+              severity: INFO
+      - extract_data:
+          assign:
+          - data: ${event.data.protoPayload}
+      - init:
+          assign:
+            - datasetsBucket: ${sys.get_env("datasets_bucket_name")}
+            - reportsBucket: ${sys.get_env("reports_bucket_name")}
+            - validatorEndpoint: ${sys.get_env("validator_endpoint")}
+            - feedID: ${text.split(data.resourceName, "/")[5]}
```
Comment (jcpitre) [inline workflows/gtfs_validator_execution.yml:17]: Would it be possible to put a comment with a real example of a resourceName? So we can understand why [5] and [6] are used.

File: workflows/gtfs_validator_execution.yml [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            - datasetID: ${text.split(data.resourceName, "/")[6]}
+            - url: ${"https://storage.googleapis.com/" + datasetsBucket + "/" + feedID + "/" + datasetID + "/" + datasetID + ".zip"}
+            - endpoint: ${validatorEndpoint + "/create-job"}
```
Comment (jcpitre) [inline workflows/gtfs_validator_execution.yml:20]: I can't see the run-validator call. Don't we need to call this to do the actual validation?

File: workflows/gtfs_validator_execution.yml [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                  steps:
+                    - known_errors:
+                        switch:
+                          - condition: ${e.code == 404}
+                            next: checkCounter
+                    - unhandled_exception:
+                        raise: ${e}
+            - checkCounter:
+                switch:
+                  - condition: ${counter < 120} # 20 minutes
```
Comment (jcpitre) [inline workflows/gtfs_validator_execution.yml:94]:
```suggestion
                 # Looping 120 times with a delay of 10 seconds each loop gives about 20 minutes. Time to quit
                  - condition: ${counter < 120} 
```

File: workflows/gtfs_validator_execution.yml [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            createJobResponse: response
+      - successCreatingJob:
+          assign:
+              - jobID: ${response.body.jobId}
+      - logJobID:
+          call: sys.log
+          args:
+              text: '${"jobID: " + jobID}'
+              severity: INFO
+      - waitForExecutionResultCreation:
```
Comment (jcpitre) [inline workflows/gtfs_validator_execution.yml:66]: My friend copilot tells me there is a retry mechanism you might be able to use.
```
- waitForExecutionResultCreation:
    retry:
      maxAttempts: 120
      delay: 10s
    steps:
      ...

```
If it does work, it looks simpler that the looping used here. Maybe you can consider it?

File: workflows/gtfs_validator_execution.yml (hunk 2 above)
Comment (cka-y) [reply workflows/gtfs_validator_execution.yml:20]: Not that I'm aware of. From reviewing the UI, we have two scenarios for handling uploads:

- **Scenario 1: User uploads a zip file.** We call an endpoint to create a job, then use another endpoint to upload the zip via the URL provided in the job creation response. The validator runs once the file is uploaded.

- **Scenario 2: User provides a URL.** This URL is directly used in the POST request to create the job, streamlining the process.

We're using Scenario 2 in this case, which directly utilizes the provided URL for job creation.

Comment (davidgamez) [reply workflows/gtfs_validator_execution.yml:20]: The validation is triggered when a file is placed in the right GCP bucket. If the URL is provided, the create-job places the downloaded file in the bucket. When the user uploads the file, the create-job returns a signed URL, and the client is responsible for uploading the file to the bucket.

File: workflows/gtfs_validator_execution.yml (hunk 4 above)
Comment (cka-y) [reply workflows/gtfs_validator_execution.yml:67]: It ended up being slightly more complicated that what your friend estimated, but it's done ✅

File: workflows/gtfs_validator_execution.yml [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+          args:
+              text: ${"JSON report copied to " + datasetsBucket + "/" + feedID + "/" + datasetID + "/report.json"}
+              severity: INFO
+          next: successfulExecution
+      - fileExistenceTimeout:
+          steps:
+            - logFileExistenceTimeout:
+                call: sys.log
+                args:
+                    text: File not found after 20 minutes
```
[the remark this answers is not in the document]
Comment (cka-y) [reply workflows/gtfs_validator_execution.yml]: I added the `datasetID` it's easier to trace using that

File: workflows/gtfs_validator_execution.yml [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            destinationObject: ${feedID + "%2F" + datasetID + "%2Freport.json"}
+            sourceBucket: ${reportsBucket}
+            sourceObject: ${jobID + "%2Freport.json"}
+            destinationPredefinedAcl: publicRead
+      - logJSONReportCopied:
+          call: sys.log
+          args:
+              text: ${"JSON report copied to " + datasetsBucket + "/" + feedID + "/" + datasetID + "/report.json"}
+              severity: INFO
+          next: successfulExecution
```
Comment (davidgamez) [inline workflows/gtfs_validator_execution.yml:146]: Let's also copy the _system_errors.json_ files that is important in case of failures.

Comment (cka-y) [reply workflows/gtfs_validator_execution.yml:160]: Done ✅ 
image

**Review by davidgamez — APPROVED** [review]
Great Work! 💪

File: workflows/gtfs_validator_execution.yml (hunk 2 above)
Comment (jcpitre) [reply workflows/gtfs_validator_execution.yml:21]: I see. It's all done with a notification from the user-uploads bucket. The pubsub subscription calls the run-validator endpoint automatically.

**Pull request merged**